src/installer/corehost/cli/apphost/bundle/dir_utils.cpp

```diff
@@ -91,3 +97,46 @@ void dir_utils_t::fixup_path_separator(pal::string_t& path)
         }
     }
 }
+
+// Retry the rename operation with some wait in between the attempts.
+// This is an attempt to workaround for possible file locking caused by AV software.
+
+bool dir_utils_t::rename_with_retries(pal::string_t& old_name, pal::string_t& new_name, bool& dir_exists)
+{
+    for (int retry_count=0; retry_count < 500; retry_count++)
+    {
+        if (pal::rename(old_name.c_str(), new_name.c_str()) == 0)
+        {
+            return true;
+        }
+        bool should_retry = errno == EACCES;
+
+        if (pal::directory_exists(new_name))
+        {
+            // Check directory_exists() on each run, because a concurrent process may have
+            // created the new_name directory.
+            //
+            // The rename() operation above fails with errono == EACCESS if 
+            // * Directory new_name already exists, or
+            // * Paths are invalid paths, or
+            // * Due to locking/permission problems.
+            // Therefore, we need to perform the directory_exists() check again.
+
+            dir_exists = true;
+            return false;
+        }
+
+        if (should_retry)
+        {
+            trace::info(_X("Retrying Rename [%s] to [%s] due to EACCES error"), old_name.c_str(), new_name.c_str());
+            pal::sleep(100);
```

Comment:
This loop will run for at most 500 times, waiting at most 50s before trying.  This is also called by commit_file(), so this can take up to 50s per file in the bundle.  It looks to me that this is a bit too much...

There might be other files in the bundle that need renaming, so maybe instead of retrying them here, add it to a queue to try later?

Maybe keep track of the total amount of time taken to rename these things?

Reply:
@lpereira  I agree that the worst case looks bad -- but it is extremely unlikely. 
The long wait time can only happen if the rename succeeds close to the 500th iteration for each file. That is, when there is a real failure with EACCESS - the extraction aborts after the first file (after 50s). 

The case when recovery is triggered is rare, and is not a path to be optimized.  If the AV blocks for a long time after writing each file, then the recovery will need to wait for it. Still, it is not expected to take the full minute after each file is written. 

While the wait time could be used to write other files that need to be recovered, the time gained is rather negligible, since the time to write these files is a few milliseconds, compared to the minute spent waiting for the AV.

I actually tried out writing the change you suggested. Please take a look [at this diff](https://github.com/swaroop-sridhar/runtime/compare/1fcheck...swaroop-sridhar:2fcheck?expand=1). But I'm not sure it is work to take in that complexity in a servicing fix, given that this case is unlikely, and that such a case is hard to test. If you still think we need to take the change, I can add it on to the PR.

Reply:
Since this is a servicing fix and this case is rare, we can keep the current behaviour from this PR.  However, keep the code from this diff somewhere, should we need it for the next servicing release (it looks pretty clean and makes the worst case scenario more manageable).

Reply:
Thanks @lpereira.
 Do you have any further comments/concerns  on this PR?